File: 1552_Magnetic_Force_Between_Two_Balls/Solution.java
import java.util.*;
class Solution {
    public boolean canPlaceBall(int x, int[] position, int m){
        int ball_placed = 1;
        int last_position = position[0];
        for (int i=1; i<position.length;i++){
            if((position[i]- last_position) >=x){
                ball_placed += 1;
                last_position = position[i];
            }
            if(ball_placed == m){
                return true;
            }
        }
        return false;
    }
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int low = 1;
        int high = position[position.length -1]/(m-1);
        int answer = 1;
        while(low<=high){
            int mid = low + (high-low)/2;
            if (canPlaceBall(mid, position, m)){
                answer = mid;
                low = mid+1;
            }
            else{
                high = mid-1;
            }

        }
        return answer;

    }
}

File: 1552_Magnetic_Force_Between_Two_Balls/Solution.py
from typing import List
class Solution:
    def canPlaceBall(self, x, position, m):
        balls = 1
        curr_position = position[0]
        for i in range(1, len(position)):
            if position[i]-curr_position >=x:
                balls += 1
                curr_position = position[i]
            if balls == m:
                return True
        return False
    def maxDistance(self, position: List[int], m: int) -> int:
        position = sorted(position)
        low = 1
        high = int(position[-1]/(m-1))
        answer = 1
        while(low<=high):
            mid = low + (high-low)/2
            mid = int(mid)
            if self.canPlaceBall(mid, position, m):
                answer= mid
                low = mid+1
            else:
                high = mid-1
        return answer

